#162192206 Intercept and handle network errors

mutaimwiti opened the pull request:
**What does this PR do?**

This PR handles network errors application-wide. This is to prevent developers from handling them everywhere that they make API calls.

**Description of Task to be completed?**

Network errors should be handled at one point to prevent repetition of code. Since the entire application uses a single `api` module that wraps `axios` to make http requests, all network errors should be handled by it. A modal should be used to indicate network errors. All areas of the application that have handled network errors should be refactored to be in line with this feature.

**How should this be manually tested?**

This can be tested by following these steps:
1. Open this PRs review app.
2. Click to open any article on the homepage. Allow it to load.
3. Go to your browser dev tools and simulate that the page is offline by checking Offline on the network tab.
4. Wait for 30 seconds and a modal will appear indicating that you're offline. The network error is automatically detected when the application tries to update article reactions and fails.
5. Go back to your browser dev tools disable offline mode on the network tab.
6. Wait for 30 seconds and the network error modal will disappear indicating that there is a connection.

**Any background context you want to provide?**

This solution is making use of `axios` interceptors, https://github.com/axios/axios#interceptors.

**What are the relevant pivotal tracker stories?**

[#162192206](https://www.pivotaltracker.com/story/show/162192206)

**Screenshots**

Simulating network error on browser dev tools.
<img width="436" alt="screen shot 2018-11-27 at 23 52 22" src="https://user-images.githubusercontent.com/8180548/49110826-93e85980-f29f-11e8-9f1c-b26b5204d3d7.png">

Viewing an article
<img width="875" alt="screen shot 2018-11-29 at 07 09 59" src="https://user-images.githubusercontent.com/8180548/49198902-6fc37000-f3a6-11e8-9220-3ad28275782d.png">

Network error occurs while viewing
<img width="872" alt="screen shot 2018-11-29 at 07 10 39" src="https://user-images.githubusercontent.com/8180548/49198913-7e118c00-f3a6-11e8-89af-d02c8e6fb87a.png">

**Questions**
Anyone that has successfully tested a materialize modal can give ideas on how we can test this.

mutaimwiti committed:
feat(nw-errors): create network error popup and integrate store

- create network error popup component
- create state to be used by network error popup
- intercept response network errors and dispatch network error
- intercept response and dispatch no network error to clear errors

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): test NetworkPopup reducer

- test that it return initial state when there is no action
- test that it handles NETWORK_ERROR_DETECTED
- test that it handles NO_NETWORK_ERROR

[Starts #162192206]

mutaimwiti committed:
feat(nw-error): test NetWorkPopup component

- test that it mounts without crashing

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): fix all dangerous uses of error.response.data

- fix like dislike
- fix login login
- fix forgot password
- fix reset password
- fix signup
- explicitly name error response on article listing
- explicitly name error response on comment thread
- explicitly name error response on comment editor

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): fix profile edit modal not working

- select profile modal by it's id instead of selecting all modals
- selecting all causes the edit profile modal not to work

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): fix like dislike linting issue

- fix Unexpected line break before closing brace

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): use class instead of id for network modal

- fix hound issue - Avoid using id selectors

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): stop setting instance on componentDidMount

- fix hound issue - Do not use setState in componentDidMount

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): use ref instead of querySelector

- use ref which is the preferred way in react

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): test network popup full dom mount

- test that it mounts on a full dom without crashing

[Starts #162192206]

paulrimiru commented on src/containers/NetworkPopup/index.js:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  onClick = () => {
+    reloadLocation();
```
Add tests for this function

mutaimwiti replied:
I have added the test @PaulKariukiRimiru

paulrimiru commented on src/containers/NetworkPopup/index.js:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  componentDidUpdate() {
+    const { networkError } = this.props;
```
Add tests for this functionality lines `25 - 30` i feel that they are important to your functionality

mutaimwiti replied:
<img width="1019" alt="screen shot 2018-11-29 at 16 35 09" src="https://user-images.githubusercontent.com/8180548/49225342-d66f7a80-f3f4-11e8-87a4-c3811cad490f.png">

@PaulKariukiRimiru an attempt to test it has proved futile. After a lot research and seeking of peer assistance I have not been able to come up with a test. See https://stackoverflow.com/questions/45833331/jest-can-not-deal-with-insertadjacentelement. This is the result of a `jsdom` shortcoming.

paulrimiru commented on src/index.js:
```diff
@@ -8,10 +8,14 @@ import * as serviceWorker from './serviceWorker';
 ReactDOM.render(
   <Provider store={store}>
-    <Router />
+    <>
```
Use `Fragments` instead

mutaimwiti replied:
It's done

mutaimwiti committed:
feat(nw-errors): test reconnect loads the current location

- test reconnect button click triggers current location reload

[Starts #162192206]

mutaimwiti committed:
feat(nw-errors): use fragment instead of <>

- use <React.Fragment> to be more explicit

[Starts #162192206]

paulrimiru committed:
Merge branch 'develop' into ft-handle-nw-errors-162192206

paulrimiru committed:
Merge branch 'develop' into ft-handle-nw-errors-162192206
